[MRG+2] DOC correct docstring for sample_gaussian

kwohlfahrt opened the pull request:
#### Reference Issue

None.
#### What does this implement/fix? Explain your changes.

Docstring for `sklearn.mixture.sample_gaussian` did not match function behaviour. Caused some trouble trying to implement a compatible version for a different distribution. Corrects this for future reference.

Commit:
DOC correct docstring for sample_gaussian

Docstring did not match function behaviour. Caused some trouble trying to implement a compatible version for a different distribution.

tguillemot commented:
@kwohlfahrt Thanks. 
Indeed there is a problem with the doc of the function, but as the GMM class is deprecated, those function will be remove at version 0.20. 
I have forgotten to put a deprecation warning to say to use make_blob instead.
Can you add the deprecation warning with your solution ?

kwohlfahrt commented:
Ah. It looks to me like the replacement does not allow for different covariances, is that correct?

tguillemot commented:
Indeed you right. I thought that `make_blob` could have any kind of covariances.
In fact, what you should use is `numpy.multivariate_normal` which is exactly the `sample_gaussian` function.

agramfort commented on sklearn/mixture/gmm.py:
```diff
@@ -64,7 +64,8 @@ def log_multivariate_normal_density(X, means, covars, covariance_type='diag'):
 
-
+@deprecated("The class function sample_gaussian is deprecated and "
+            "will be removed in 0.20. Use numpy.multivariate_normal instead.")
```
if you deprecate a function you need to remove it from everywhere. is this used somewhere? otherwsie the test suite will throw a lot of deprecation warnings

kwohlfahrt replied:
Looking at the logs, there is only one test which uses it, `test_sample_gaussian`.

tguillemot replied:
If you do git grep sample_gaussian, you'll see that the class _GMM uses it.
So we need to create a _sample_gaussian function.

agramfort commented on sklearn/mixture/gmm.py:
```diff
@@ -64,7 +64,8 @@ def log_multivariate_normal_density(X, means, covars, covariance_type='diag'):
         X, means, covars)
 
-
+@deprecated("The class function sample_gaussian is deprecated and "
```
The class function -> Function

tguillemot commented:
@agramfort Sorry I've forgot that.
What I suggest is to create a private function with the name `_sample_gaussian` and create another public function call `sample_gaussian` which calls it. We have done a similar thing with the GMM class.

It will solve any problems until we remove completely the old GMM classes.

agramfort commented:
sounds like a plan

tguillemot commented:
>Looking at the logs, there is only one test which uses it, test_sample_gaussian

With `git grep sample_gaussian`, it appears that the `_GMM` class uses `sample_gaussian` in `sample`.
So we need to create a `_sample_gaussian` function.

kwohlfahrt commented:
OK. Should the tests that test `sample_gaussian` directly be moved to the new function as well?

agramfort commented:
Yes






On Thu, Jul 7, 2016 at 7:14 PM +0200, "Kai" <notifications@github.com> wrote:












OK. Should the tests that test sample_gaussian directly be moved to the new function as well?



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

Commit:
MAINT Use _sample_gaussian internally

This avoids triggering the deprecation warning. sample_gaussian is just a thin
wrapper.

agramfort commented on sklearn/mixture/gmm.py:
```diff
@@ -64,7 +64,8 @@ def log_multivariate_normal_density(X, means, covars, covariance_type='diag'):
 
-
+@deprecated("The function sample_gaussian is deprecated and "
+            "will be removed in 0.20. Use numpy.multivariate_normal instead.")
```
I suppose it's numpy.random.multivariate_normal

tguillemot replied:
Indeed it's `numpy.random.multivariate_normal`.

agramfort commented:
CIs are not happy

tguillemot commented:
You have three other things to change in `test_gmm.py`:
- add the import `from sklearn.mixture.gmm import _sample_gaussian`
- remove all the `mixture.` before sample_gaussian
- change the comment l26  `#Test sample generation from mixture.sample_gaussian...` to `... gmm._sample_gaussian...`
- remove the import l59 `from sklearn.mixture import sample_gaussian`

tguillemot commented:
I think the appveyor failure it's unrelated but to be sure can you do ``git commit --amend; git push -f``.

amueller commented on sklearn/mixture/gmm.py:
```diff
@@ -64,7 +64,9 @@ def log_multivariate_normal_density(X, means, covars, covariance_type='diag'):
         X, means, covars)
 
-
+@deprecated("The function sample_gaussian is deprecated and "
```
Please add the version they were deprecated in to the message (was deprecated in 0.18)

amueller commented:
LGTM apart from adding the deprecation version. This was not superseeded by #7012, was it?

tguillemot commented:
@amueller No as @kwohlfahrt was working on it I didn't want to remove his work.
@kwohlfahrt Can you fix the comments ?

Commit:
Fix _sample_gaussian test path

_sample_gaussian is not publicly exported in mixture, so use gmm module.

kwohlfahrt commented:
Hi all, sorry I dropped this for so long. Rebased and fixed the tests, so if you still want to merge go ahead!

amueller commented:
lgtm. @tguillemot ?

amueller commented:
@tguillemot good to merge?

tguillemot commented:
LGTM. Sorry for the delay.

TomDLT commented:
Thanks @kwohlfahrt
